### PR TITLE
statistics: migrate test-infra to testify for statistics/handle/gc_test.go

### LINKED ISSUE #28128: migrate test-infra to testify for `statistics/handle/gc_test.go`

### PR DESCRIPTION
**Pull request by feitian124**
### What problem does this PR solve?

Issue Number: close #28128

### Release note
```release-note
None
```

### CODE REVIEW

**Comment by feitian124** [comment]
have problems, still work in progress

**Comment by feitian124** [comment]
after migrate to testify, below test is error:

```go
	testKit.MustQuery("select count(*) from mysql.stats_histograms").Check(testkit.Rows("4"))
	testKit.MustQuery("select count(*) from mysql.stats_buckets").Check(testkit.Rows("12"))
```

```
    result.go:49: 
        	Error Trace:	result.go:49
        	            				gc_test.go:67
        	Error:      	Not equal: 
        	            	expected: "[12]\n"
        	            	actual  : "[0]\n"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-[12]
        	            	+[0]
        	            	 
        	Test:       	TestGCStats
        	Messages:   	sql:select count(*) from mysql.stats_buckets, args:[]
```

i tested in tidb playground 5.2.1, stats_histograms has 4 records, stats_buckets has 0 records.
seems the orgin `TestGCStats` is not correct or not run.

out put of `select * from mysql.stats_histograms `:

```
#,table_id,is_index,hist_id,distinct_count,null_count,tot_col_size,modify_count,version,cm_sketch,stats_ver,flag,correlation,last_analyze_pos
1,55,0,1,3,0,27,0,427971302335709192,,2,1,1,
2,55,0,2,3,0,27,0,427971302335709192,,2,1,1,
3,55,1,1,3,0,54,0,427971302335709192,,2,1,0,
4,55,1,2,3,0,27,0,427971302335709192,,2,1,0,
```

out put of `select * from mysql.stats_buckets `:

```

```

File: statistics/handle/gc_test.go [hunk 1]
```diff
@@ -15,16 +15,48 @@
...
 	"github.com/pingcap/tidb/sessionctx/variable"
-	"github.com/pingcap/tidb/util/testkit"
+	"github.com/pingcap/tidb/testkit"
+	"github.com/stretchr/testify/require"
 )
 
-func (s *testStatsSuite) TestGCStats(c *C) {
-	defer cleanEnv(c, s.store, s.do)
-	testKit := testkit.NewTestKit(c, s.store)
+func createTestKitAndDom(t *testing.T) (*testkit.TestKit, *domain.Domain, func()) {
```
Comment (tisonkun) [inline statistics/handle/gc_test.go:28]: If you create new store for each test case, I think you don't have to implement `statsClean`.

Comment (feitian124) [reply statistics/handle/gc_test.go:28]: ok, get it.
the current problem is that i can not find the reason why it fails after migrate, need more time.

Comment (tisonkun) [reply statistics/handle/gc_test.go:28]: I'm trying to run the test before migrated, it gives

* failed if run only the test
* success if run the whole package

I think there must be some share state in the whole package. Maybe @winoros has more ideas on it.

Comment (feitian124) [reply statistics/handle/gc_test.go:28]: thank you for help checking, @tisonkun . 
i checked carefully myself and believe i only did simple one - one replacement migrate, not sure why it fails and have to dig more to find the reason.
although this is a good chance to understand the logic of statistics(actually i read a lot), guide to how to locate the reason is still appreciate。@winoros

**Comment by xuyifangreeneyes** [comment]
@feitian124 I think that the expect result of `select count(*) from mysql.stats_buckets` is `0` rather than `12` if we only run `TestGCStats` without running other tests(I ran the single test `TestGCStats` for master branch on my computer and it failed). The default value of `tidb_analyze_version` is `2` and in analyze version 2, histogram don't contain topN. In another word, histogram and topN together represent all the samples. After executing the following SQLs,
```
create table t(a int, b int, index idx(a, b), index idx_a(a))
insert into t values (1,1),(2,2),(3,3)
analyze table t
```
The row number is too small so that all the samples are in topN and no sample is in buckets of histogram. You can check `select * from mysql.stats_top_n`, there are 12 rows and each of `a`,`b`,`idx`,`idx_a` has 3 respectively. On the other hand, there is no row for `select * from mysql.stats_buckets`.

If the analyze version is 1(`tidb_analyze_version=1`), histogram represent all the samples so the result of `select count(*) from mysql.stats_buckets` is `12`.

I guess that the reason of `TestGCStats` not failing before may be that some other test modifies `tidb_analyze_version` to `1`. IMO, you can add `set @@tidb_analyze_version = 1` to make the pr pass `TestGCStats`.

**Comment by xuyifangreeneyes** [comment]
@feitian124 I think that the expect result of `select count(*) from mysql.stats_buckets` is `0` rather than `12` if we only run `TestGCStats` without running other tests(I ran the single test `TestGCStats` for master branch on my computer and it failed). The default value of `tidb_analyze_version` is `2` and in analyze version 2, histogram don't contain topN. In another word, histogram and topN together represent all the samples. After executing the following SQLs,
```
create table t(a int, b int, index idx(a, b), index idx_a(a))
insert into t values (1,1),(2,2),(3,3)
analyze table t
```
The row number is too small so that all the samples are in topN and no sample is in buckets of histogram. You can check `select * from mysql.stats_top_n`, there are 12 rows and each of `a`,`b`,`idx`,`idx_a` has 3 respectively. On the other hand, there is no row for `select * from mysql.stats_buckets`.

If the analyze version is 1(`tidb_analyze_version=1`), histogram represent all the samples so the result of `select count(*) from mysql.stats_buckets` is `12`.

I guess that the reason of `TestGCStats` not failing before may be that some other test modifies `tidb_analyze_version` to `1`. IMO, you can add `set @@tidb_analyze_version = 1` to make the pr pass `TestGCStats`.

File: statistics/handle/gc_test.go [hunk 2]
```diff
@@ -15,16 +15,47 @@
...
+		dom.StatsHandle().Clear()
+		clean()
+	}
+	return tk, dom, statsClean
+}
+
+func TestGCStats(t *testing.T) {
+	testKit, dom, clean := createTestKitAndDom(t)
+	defer clean()
 	testKit.MustExec("use test")
```
Comment (xuyifangreeneyes) [inline statistics/handle/gc_test.go:59]: Try to add `testKit.MustExec("set @@tidb_analyze_version = 1")` here.

**Comment by feitian124** [comment]
> The default value of `tidb_analyze_version` is `2` and in analyze version 2, histogram don't contain topN. In another word, histogram and topN together represent all the samples.

@xuyifangreeneyes understand. thank you for your detail explain, this helps a lot.

so some tests in package statistics is for test tidb_analyze_version = 2, some is for  tidb_analyze_version = 2, and maybe some works on both.

when migrate remaining tests of this package, i need pay attention to this.

**Comment by feitian124** [comment]
i think this pr has been down, please help review, thanks.

/cc @tisonkun @xuyifangreeneyes

**Review by tisonkun — APPROVED** [review]

**Comment by xuyifangreeneyes** [comment]
> so some tests in package statistics is for test tidb_analyze_version = 1, some is for tidb_analyze_version = 2, and maybe some works on both.

Yes. Feel free to ask us if tidb_analyze_version of some tests is unclear.

**Review by xuyifangreeneyes — APPROVED** [review]

**Comment by feitian124** [comment]
@mmyj could you help merge? thanks

**Pull request merged**